File: Programmers/2/142085. 디펜스 게임/디펜스 게임.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(int n, int k, vector<int> enemy) {
    priority_queue<int> pq;

    for (int i = 0; i < (int)enemy.size(); i++)
    {
        n -= enemy[i];
        pq.push(enemy[i]);

        if (n < 0)
        {
            if (k <= 0)
                return i;
            
            int largest = pq.top(); pq.pop();
            n += largest;
            k--;
            
            if (n < 0)
                return i;
        }
    }
    
    return (int) enemy.size();
}

File: Programmers/2/142085. 디펜스 게임/디펜스 게임.java
import java.util.Collections;
import java.util.PriorityQueue;

class Solution {
    public int solution(int n, int k, int[] enemy) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
                
        for (int i = 0; i < enemy.length; i++) {
            n -= enemy[i];
            maxHeap.offer(enemy[i]);
            
            if (n < 0) {
                if (k <= 0)
                    return i;
                
                int largest = maxHeap.poll();
                n += largest;
                k--;
                
                if (n < 0)
                    return i;
            }
        }
        
        return enemy.length;
    }
}